src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +114,162 @@ internal class DiagnosticComputer
+            //     before starting the compute.
+            //   - Suspended normal priority requests are re-attempted in the below loop.
+            bool suspended;
+
+            do
+            {
+                suspended = false;
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Create a linked cancellation source to allow high priority tasks to cancel normal priority tasks.
+                using var cancellationTokenSource = new CancellationTokenSource();
+                using var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationTokenSource.Token, cancellationToken);
+
+                // Fetch the cancellation token here to avoid capturing linkedCts in the computeTask lambda as the task may run after linkedCts has been disposed due to cancellation.
+                var linkedCancellationToken = linkedCancellationTokenSource.Token;
+
+                await WaitOrSuspendExecutingTasksAsync(highPriority, cancellationToken).ConfigureAwait(false);
+
+                var computeTask = Task.Run(async () =>
+                {
+                    var diagnosticsComputer = new DiagnosticComputer(document, project,
+                        solutionChecksum, ideOptions, span, analysisKind, analyzerInfoCache, hostWorkspaceServices);
+                    return await diagnosticsComputer.GetDiagnosticsAsync(analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, linkedCancellationToken).ConfigureAwait(false);
+                }, linkedCancellationToken);
+
+                StartTrackingPreCompute(computeTask, cancellationTokenSource, highPriority);
+
+                try
+                {
+                    return await computeTask.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    if (!cancellationTokenSource.IsCancellationRequested)
+                    {
+                        throw;
+                    }
+
+                    // Normal priority task that was suspended by a high priority task.
+                    Debug.Assert(!highPriority);
+                    suspended = true;
+                }
+                finally
+                {
+                    StopTrackingPostCompute(computeTask, cancellationTokenSource, highPriority);
+                }
+            } while (suspended);
+
+            throw ExceptionUtilities.Unreachable();
+
+            static async Task WaitOrSuspendExecutingTasksAsync(bool highPriority, CancellationToken cancellationToken)
+            {
+                // High priority task forces suspension of all the executing normal priority tasks.
+                // Normal priority task waits for all the executing high priority tasks to complete.
+                if (highPriority)
```

Comment:
nit: looking through, we basically have two separate codepaths.  one for high-pri and one for normal-pri.  And within these bodies, theya re not sharing logic.  I think teh switch on this shoudl be the first thing that happens, and then we have the dedicated logic for the high-pri work written out and the dedicated logic for the normal-pri work.  That will make it much easier to understand IMO as we won't be overlapping all this logic.

Reply:
> I think teh switch on this shoudl be the first thing that happens

I thought about this and the code becomes harder to maintain. All the local functions have analogous logic for high priority and normal priority requests, just operating on different data structures. Separating out these code paths upfront will duplicate the algorithm, and leaves us susceptible for any future fixes or refactorings in one path not to be applied to the other. Current implementation safeguards us from this. I prefer staying with the current design.